Objects/typeobject.c

```diff
@@ -4220,8 +4220,8 @@ _PyType_LookupId(PyTypeObject *type, _Py_Identifier *name)
 
 /* This is similar to PyObject_GenericGetAttr(),
    but uses _PyType_Lookup() instead of just looking in type->tp_dict. */
-static PyObject *
-type_getattro(PyTypeObject *type, PyObject *name)
+PyObject *
+_Py_type_getattro_impl(PyTypeObject *type, PyObject *name, int *suppress)
```

Comment:
I don't really like the design of this. Why not pass in `int supress` instead of `int *supress` ? Then it's just a simple boolean flag. For the caller (`_PyObject_LookupAttr`) we just need to check that `result == NULL && supresss` then return 0.

Reply:
I was not really happy with the pointer either. Your suggestion is more elegant, so I implemented it in https://github.com/python/cpython/compare/main...eendebakpt:cpython:type_hasattr_v2. The build fails in the deepfreeze stage. I suspect this is due to the following:

* In the `_Py_type_getattro_impl` method there are multiple paths that can return a NULL value. For example in `_Py_type_getattro_impl`, this is one such path
```
    /* Initialize this type (we'll assume the metatype is initialized) */
    if (!_PyType_IsReady(type)) {
        if (PyType_Ready(type) < 0)
            return NULL;
    }
```
(there are a few more paths)

* In case of a missing attribute the `PyExc_AttributeError` is set (and suppressed), but for the other paths another error can be set. 
Then in `_PyObject_LookupAttr` from `object.c` this code
```
    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
        return -1;
    }
```
returns a -1, where in the [type_hasattr_v2](https://github.com/python/cpython/compare/main...eendebakpt:cpython:type_hasattr_v2) branch we return 0.


I do not think we can safely use any value other than NULL as a return value from `_Py_type_getattro_impl` to signal the missing attribute case (any other value can be a valid address in memory).

So I am not sure how to avoid the pointer argument in an elegant way.

@Fidget-Spinner

Reply:
OK so just to be clear, the `supress_missing_attribute_exception` flag only has two values:
```
NULL - don't supress
any int (including 0) - supress
```
and it's not like the other flags where 0 is false and 1 is true, right?

Reply:
The `supress_missing_attribute_exception ` is indeed different from a simple boolean flag. It has only two values, but it also returns a value:

* `NULL`: do not supress the exception (original behaviour)
* Non-zero: suppress the `PyExc_AttributeError` and set  `*supress_missing_attribute_exception` to 1 to signal we have suppressed the exception and are not in one of the other code paths.

We could modify this to:
```
_Py_type_getattro_impl(PyTypeObject *type, PyObject *name, int &raise_exception)
```
With the following two modes:

*  `raise_exception=1`: original behaviour (do not suppress)
* `raise_exception=0`: all code paths identical, except for the `PyExc_AttributeError` path: there we suppress the `PyExc_AttributeError` and set `raise_exception=1` to signal the caller we have suppressed the missing attribute error

This is more or less the same as the current PR, but I like this option better since we are now setting a reference instead of a pointer. @Fidget-Spinner

Reply:
I think the current way is fine, but we should document this somewhere. Could you add a comment above the function definition explaining the values of `supress_missing_attribute_exception` ? Copy pasting this is fine:

> * `NULL`: do not suppress the exception
> * Non-zero: suppress the `PyExc_AttributeError` and set  `*supress_missing_attribute_exception` to 1 to signal we have suppressed the exception and are not in one of the other code paths.